central/networkgraph/entity/datastore/datastore_impl_sac_test.go

```diff
@@ -0,0 +1,760 @@
+			expectedError: nil,
+		},
+		"Partial cluster read/write cannot create a global entity": {
+			contextName:   sacTestUtils.Cluster1NamespaceAReadWriteCtx,
+			expectedError: sac.ErrResourceAccessDenied,
+		},
+		"Other cluster read/write cannot create a cluster entity for cluster not in scope": {
+			contextName:   sacTestUtils.Cluster2ReadWriteCtx,
+			expectedError: sac.ErrResourceAccessDenied,
+		},
+	}
+}
+
+func (s *NetworkEntityDataStoreSACTestSuite) TestExistsSAC() {
+	s.setupSACReadTest()
+	testContexts := sacTestUtils.GetNamespaceScopedTestContexts(context.Background(), s.T(), resources.NetworkGraph)
+
+	for name, tc := range getReadTestCases(s.T()) {
+		s.Run(name, func() {
+			ctx := testContexts[tc.contextName]
+			exists, testErr := s.ds.Exists(ctx, tc.entityID)
+			s.NoError(testErr)
+			s.Equal(tc.expectedFound, exists)
+		})
+	}
+}
+
+func (s *NetworkEntityDataStoreSACTestSuite) TestGetIDsSAC() {
+	s.setupSACReadTest()
+	testContexts := sacTestUtils.GetNamespaceScopedTestContexts(context.Background(), s.T(), resources.NetworkGraph)
+
+	for name, tc := range getListTestCases(s.T()) {
+		s.Run(name, func() {
+			ctx := testContexts[tc.contextName]
+			fetchedIDs, testErr := s.ds.GetIDs(ctx)
+			s.NoError(testErr)
+			s.ElementsMatch(tc.expectedIDs, fetchedIDs)
+		})
+	}
+}
+
+func (s *NetworkEntityDataStoreSACTestSuite) TestGetEntitySAC() {
+	s.setupSACReadTest()
+	testContexts := sacTestUtils.GetNamespaceScopedTestContexts(context.Background(), s.T(), resources.NetworkGraph)
+
+	for name, tc := range getReadTestCases(s.T()) {
+		s.Run(name, func() {
+			ctx := testContexts[tc.contextName]
+			entity, found, testErr := s.ds.GetEntity(ctx, tc.entityID)
+			s.NoError(testErr)
+			if tc.expectedFound {
+				s.True(found)
+				protoassert.Equal(s.T(), tc.expectedEntity, entity)
+			} else {
+				s.False(found)
+				s.Nil(entity)
+			}
+		})
+	}
+}
+
+func (s *NetworkEntityDataStoreSACTestSuite) TestGetEntityByQuerySAC() {
+	// Note: The test here reflects the observed behavior,
+	// which is not the expected one from a pure scoped access control perspective
+	// (entities that are not linked to the requester scope should be filtered out,
+	// like global entities for users that have access scope restricted to a cluster
+	// or cluster entities for clusters that are not in the access scope of the user).
+	s.setupSACReadTest()
+	testContexts := sacTestUtils.GetNamespaceScopedTestContexts(context.Background(), s.T(), resources.NetworkGraph)
+
+	globalEntityID := getGlobalEntityID(s.T())
+	cluster1EntityID := getCluster1EntityID(s.T())
+	missingEntityID := getCluster2EntityID(s.T())
+	allEntityIDs := []string{globalEntityID.String(), cluster1EntityID.String(), missingEntityID.String()}
+
+	globalEntity := getGlobalEntity(s.T())
+	clusterEntity := getCluster1Entity(s.T())
+	allEntities := []*storage.NetworkEntity{globalEntity, clusterEntity}
+	// clusterEntities := []*storage.NetworkEntity{clusterEntity}
```

Comment:
Why it's commented out? Do we need it?

Reply:
The point here is to highlight the difference between the test behaviour and the expected one.
The current implementation of GetEntityByQuery does not perform much access control checks. The commented test code or test descriptions highlight what the actual function behaviour could or should be.